docs/source/getting-started/testing.rst

```diff
@@ -0,0 +1,118 @@
+    - We use the ``@daml/ledger`` library to check the ledger state. In this case, we want to ensure there is a single ``User`` contract created for the new party. Hence we create a new connection to the ``Ledger``, ``query()`` it and state what we ``expect`` of the result. When we run the tests, Jest will check these expectations and report any failures for us to fix.
+    - The test also simulates the new user logging out and then logging back in. We again check the state of the ledger and see that it's the same as before.
+    - Finally we must ``close()`` the browser page, which was opened in ``newUiPage()``, to avoid runaway Puppeteer processes after the tests finish.
+
+You will likely use ``test``, ``getParty()``, ``newUiPage()`` and ``Browser.close()`` for all your tests.
+In this case we use the ``@daml/ledger`` library to inspect the state of the ledger, but usually we just check the contents of the web page match our expectations.
+
+
+Accessing UI elements
+=====================
+
+We showed how to write a simple test at a high level, but haven't shown how to make individual actions in the app using Puppeteer.
+This was hidden in the ``login()`` and ``logout()`` functions.
+Let's see how ``login()`` is implemented.
+
+.. literalinclude:: code/ui-before/index.test.tsx
+  :language: tsx
+  :start-after: // BEGIN_LOGIN_FUNCTION
+  :end-before: // END_LOGIN_FUNCTION
+
+We first wait to receive a handle to the username input element.
+This is important to ensure the page and relevant elements are loaded by the time we try to act on them.
+We then use the element handle to click into the input and type the party name.
+Next we click the login button (this time assuming the button has loaded along with the rest of the page).
+Finally, we wait until we find we've reached the menu on the main page.
+
+The strings used to find UI elements, e.g. ``'.test-select-username-field'`` and ``'.test-select-login-button'``, are `CSS Selectors <https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Selectors>`_.
+You may have seen them before in CSS styling of web pages.
+In this case we use *class selectors*, which look for CSS classes we've given to elements in our React components.
+
+This means we must manually add classes to the components we want to test.
+For example, here is a snippet of the ``LoginScreen`` React component with classes added to the ``Form`` elements.
+
+.. literalinclude:: code/ui-before/LoginScreen.tsx
+  :language: tsx
+  :start-after: // BEGIN_FORM
+  :end-before: // END_FORM
+
+You can see the ``className`` attributes in the ``Input`` and ``Button``, which we select in the ``login()`` function.
+Note that you can use other features of an element in your selector, such as its type and attributes.
+We've only used class selectors in these tests.
+
+
+Writing CSS Selectors
```

Comment:
I'm not convinced this section adds more value than confusion. I think the issues we're facing here are pretty universal for any React UI framework. Since we're not trying to teach web development but rather want to focus on the issues related to DAML application, I'd prefer to entirely remove this section, particularly since there's no actual issue but only an anticipated potential one.

Reply:
I'm not sure. Agree we're not trying to teach web dev in its entirety, but we're also not assuming the reader is familiar with these concepts. I don't think it's fair to expect the reader to automatically understand the fact that we are using Semantic UI (we actually don't mention that anywhere) which translates into HTML, and we are selecting the resulting HTML elements. This was an actual issue for me when I was writing selectors, until I realised this point. I take your point that it's in the weeds, but might still be helpful.

Reply:
Don't think this section hurts. It's short and points out a specific thing for Semantic UI when testing DAML applications built with create-daml-app (that uses the Semantic UI).